xcdat/temporal_avg.py

```diff
@@ -0,0 +1,988 @@
+        Parameters
+        ----------
+        df_season : pd.DataFrame
+            The DataFrame of time coordinates for the "season" frequency with
+            obsolete columns.
+
+        Returns
+        -------
+        pd.DataFrame
+            The DataFrame of time coordinates for the "season" frequency with
+            obsolete columns dropped.
+        """
+        if self._mode == "timeseries":
+            df_season = df_season.drop("month", axis=1)
+        elif self._mode in ["climatology", "departure"]:
+            df_season = df_season.drop(["year", "month"], axis=1)
+        else:
+            raise ValueError(
+                "Unable to drop columns in the datetime components "
+                f"DataFrame for unsupported mode, '{self._mode}'."
+            )
+
+        return df_season
+
+    def _get_weights(self, data_var: xr.DataArray) -> xr.DataArray:
+        """Calculates weights for a data variable using time bounds.
+
+        This method gets the length of time for each coordinate point by using
+        the difference in the upper and lower time bounds. This approach ensures
+        that the correct time lengths are calculated regardless of how time
+        coordinates are recorded (e.g., monthly, daily, hourly) and the calendar
+        type used.
+
+        The time lengths are grouped by the time MultiIndex, then each time
+        length is divided by the total sum of the time lengths in its group to
+        get the weights. The sum of the weights for each group is validated to
+        ensure it equals 1.0  or 0.0. A sum of 0.0 means that all values in the
+        group are np.nan (e.g., dropping incomplete "DJF" seasons).
+
+        Parameters
+        -------
+        data_var : xr.DataArray
+            The data variable.
+
+        Returns
+        -------
+        xr.DataArray
+            The weights based on a specified frequency.
+
+        Notes
+        -----
+        Refer to [5]_ for the supported CF convention calendar types.
+
+        References
+        -----
+        .. [5] https://cfconventions.org/cf-conventions/cf-conventions.html#calendar
+        """
+        time_lengths = self._time_bounds[:, 1] - self._time_bounds[:, 0]
+
+        # Attributes are dropped from time bounds DataArray after the
+        # subtraction operation, so they must be restored.
+        time_lengths.attrs = self._time_bounds.attrs
+        time_lengths[self._time_key].attrs = data_var[self._time_key].attrs
+
+        time_grouped = self._groupby_multiindex(time_lengths)
+        weights: xr.DataArray = time_grouped / time_grouped.sum()  # type: ignore
+
+        self._validate_weights(data_var, weights)
+        return weights
+
+    def _validate_weights(self, data_var: xr.DataArray, weights: xr.DataArray):
+        """Validates the sums of the weights for each group equals 1.0 or 0.0.
+
+        A sum of 0.0 means that all values in the group are np.nan (e.g.,
+        dropping incomplete "DJF" seasons).
```

Comment:
I wasn't sure if this case can occur, since DJF is dropped from the DataArray.

Reply:
Thanks for catching this incorrect description. DJF seasons should be dropped prior to calculating weights, rather than masked with `np.nan`.

I incorrectly wrote two DJF tests for `_calculate_weights()` that had `np.nans` representing the dropped seasons (they should be removed instead of masked).